ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java

```diff
@@ -0,0 +1,242 @@
+  private final ProtocolSchedule protocolSchedule;
+  private final BlockchainQueries blockchainQueries;
+  private final Blockchain blockchain;
+
+  public EthFeeHistory(
+      final ProtocolSchedule protocolSchedule, final BlockchainQueries blockchainQueries) {
+    this.protocolSchedule = protocolSchedule;
+    this.blockchainQueries = blockchainQueries;
+    this.blockchain = blockchainQueries.getBlockchain();
+  }
+
+  @Override
+  public String getName() {
+    return RpcMethod.ETH_FEE_HISTORY.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext request) {
+    final long blockCount = request.getRequiredParameter(0, Long.class);
+    final BlockParameter highestBlock = request.getRequiredParameter(1, BlockParameter.class);
+    final Optional<List<Double>> maybeRewardPercentiles =
+        request.getOptionalParameter(2, Double[].class).map(Arrays::asList);
+
+    final long chainHeadBlockNumber = blockchain.getChainHeadBlockNumber();
+    final long resolvedHighestBlockNumber =
+        highestBlock
+            .getNumber()
+            .orElse(
+                highestBlock.isEarliest()
+                    ? BlockHeader.GENESIS_BLOCK_NUMBER
+                    : chainHeadBlockNumber /* both latest and pending use the head block until we have pending block support */);
+
+    final Object requestId = request.getRequest().getId();
+    if (resolvedHighestBlockNumber > chainHeadBlockNumber) {
+      return new JsonRpcErrorResponse(requestId, JsonRpcError.INVALID_PARAMS);
+    }
+
+    final long oldestBlock = Math.max(0, resolvedHighestBlockNumber - (blockCount - 1));
+
+    final List<BlockHeader> blockHeaders =
+        LongStream.range(oldestBlock, oldestBlock + blockCount)
+            .mapToObj(blockchain::getBlockHeader)
+            .flatMap(Optional::stream)
+            .collect(toUnmodifiableList());
+
+    // we return the base fees for the blocks requested and 1 more because we can always compute it
+    final List<Long> explicitlyRequestedBaseFees =
+        LongStream.range(oldestBlock, oldestBlock + blockCount)
+            .mapToObj(blockchain::getBlockHeader)
+            .map(
+                maybeBlockHeader ->
+                    maybeBlockHeader.flatMap(ProcessableBlockHeader::getBaseFee).orElse(0L))
+            .collect(toUnmodifiableList());
+    final long nextBlockNumber = resolvedHighestBlockNumber + 1;
+    final Long nextBaseFee =
+        blockchain
+            .getBlockHeader(nextBlockNumber)
+            .map(blockHeader -> blockHeader.getBaseFee().orElse(0L))
+            .orElseGet(
+                () ->
+                    protocolSchedule
+                        .getByBlockNumber(nextBlockNumber)
+                        .getEip1559()
+                        .map(
+                            eip1559 -> {
+                              final BlockHeader lastBlockHeader =
+                                  blockHeaders.get(blockHeaders.size() - 1);
```

Comment:
I think we can have a out of bound exception if blockCount==0

Maybe we need to add a check to avoid blockcount<1